### DISCUSSION TITLE
Contextual menu rearrangement

### DISCUSSION THREAD

**Discussion by pgilfernandez** [friction2d/friction#302] [Ideas]
Maybe it's because I'm still new with Friction but I find the contextual menu (secondary button menu) a little bit messy, I mean, not all options available and ordered and some "hidden" under "actions"...
What would you think to make it simpler and listed as follows?:

<img width="344" alt="menus" src="https://github.com/user-attachments/assets/1327a188-e83e-40b4-9f64-d8295ea949c3">

Arranging on top the most common or used commands makes it easier to find and use them.

The only thing that I don't know if could be done is turning the all effects types checkable and at the same time be able to open a secondary menu with their child available effects.

I would even suggest that the last 6 "effects" items loose the word "Effect" and add a title called "Effect" on top of their "group"... This way is would be simpler and easier to "scrub" with your eyes as simpler words are easier to find.

It's just an idea, what do you think?

**Comment by rodlie** [comment]
The context menu is not the best regarding UX and could be improved, but there is also a reason things are the way they are.

The "Actions" menu is the main menu, every "box" (in Friction every item in a project is a box) includes this, and it's own actions (that differs from box to box). Adding duplicates from "Actions" is IMHO a bad thing, we now have duplication of code etc. I have made exceptions in the past to include some "common" actions in the box menu, but it's still code duplication and I don't like it.

A menu refactor is probably the best option, but this will take time away from other things (features/fixes).

**Reply by rodlie** [reply]
A better way to look at it: how do we visualize a QMenu + a sub QMenu in one "menu" (widget).

**Reply by rodlie** [reply]
How context menu works in the canvas:

https://github.com/friction2d/friction/blob/main/src/core/canvasmouseinteractions.cpp#L152

Will trigger all selected boxes to generate their own menu and then is merged into one.

https://github.com/search?q=repo%3Afriction2d%2Ffriction%20setupCanvasMenu&type=code

The context menu for boxes in properties are generated here: https://github.com/friction2d/friction/blob/main/src/app/GUI/BoxesList/boxsinglewidget.cpp#L664

that triggers https://github.com/search?q=repo%3Afriction2d%2Ffriction%20prp_setupTreeViewMenu&type=code 

As you can see we don't actually know what will be added to the menu.

**Reply by rodlie** [reply]
Also a note.

An ImageBox will generate menu actions specific to images/raster stuff (and also assets actions), but also since it's a override of an BoundingBox we also need the menu actions from that... etc. The Friction Core is very modular/flexible, but with that comes UX challenges.

**Reply by pgilfernandez** [reply]
I understand, I only had the UX perspective and didn't know if it was easy/difficult/impossible to do and, more important, if it worth it or if it had sense...

**Reply by rodlie** [reply]
I'm all for fixing this, but we need to figure out how we are going to visualize it, I can override the QMenu and paint whatever.

**Reply by pgilfernandez** [reply]
I'll have a deeper look at the code you shared but I'm not very good at C++... I feel more comfortable on the graphics/design side 😅

**Reply by rodlie** [reply]
The code is not that important. When the user right-click we must show a QMenu, that QMenu has the actions associated with the current box first, then various other actions are added under, sometimes in submenus.

We could then try to figure out how we are going to visualize this, suggestions are welcome.

We also don't need to show a QMenu, could be a custom widget that does something with the menus we have.

Also see https://github.com/friction2d/friction/issues/169

**Reply by pgilfernandez** [reply]
OK, I'll have a look at it and try to suggest something doable
